engine/input.cpp

```diff
@@ -11,14 +11,47 @@ static constexpr int WINDOW_WIDTH = 1280;
 static constexpr int WINDOW_HEIGHT = 960;
 
 // All keys are intalized to false, for not pressed
-std::array<bool, GLFW_KEY_LAST + 1> Input::mKeys {};
-std::array<bool, GLFW_MOUSE_BUTTON_LAST + 1> Input::mMouseButtons {};
+std::array<InputState, GLFW_KEY_LAST + 1> Input::mKeys {};
+std::array<InputState, GLFW_MOUSE_BUTTON_LAST + 1> Input::mMouseButtons {};
 // Initalize way outside the screen so it does not trigger anything before it should
 glm::vec3 Input::mMousePosition { std::numeric_limits<float>::max() };
 
-bool Input::IsKeyPressed(unsigned int keycode)
+void Input::Init(GLFWwindow* window)
 {
-    return mKeys[keycode];
+    for (auto& inputVal : mKeys)
+    {
+        inputVal = InputState::NotPressed;
+    }
+
+    for (auto& inputVal : mMouseButtons)
+    {
+        inputVal = InputState::NotPressed;
+    }
+
+    glfwSetKeyCallback(window, Input::KeyCallback);
+    glfwSetCursorPosCallback(window, Input::MousePositionCallback);
+    glfwSetMouseButtonCallback(window, Input::MouseButtonCallback);
+}
+
+void Input::Update()
+{
+    const auto updateInputState = [](InputState* inputStates, size_t size)
```

Comment:
I don't suppose a good way to take these as a std::array is there? I looked into it, and it seems starting with C++20 you could do a templated lambda function. But there doesn't seem to be a good option otherwise
```
auto lambda = []<typename T>(T t){
};
```

Reply:
Yeah, I actually think these are templates under the hood, like if you look at what the compiler generates it's just a template. Funny they took this approach first...